src/cultionet/callbacks.py

```diff
@@ -0,0 +1,163 @@
+import typing as T
+import filelock
+from pathlib import Path
+
+import geowombat as gw
+import rasterio as rio
+from rasterio.windows import Window
+import torch
+from pytorch_lightning.callbacks import BasePredictionWriter
+from torch_geometric.data import Data
+
+from .data.const import SCALE_FACTOR
+from .utils.reshape import ModelOutputs
+
+
+class LightningGTiffWriter(BasePredictionWriter):
+    def __init__(
+        self,
+        reference_image: Path,
+        out_path: Path,
+        num_classes: int,
+        ref_res: float,
+        resampling,
+        compression: str,
+        write_interval: str = 'batch'
+    ):
+        super().__init__(write_interval)
+
+        self.reference_image = reference_image
+        self.out_path = out_path
+        self.out_path.parent.mkdir(parents=True, exist_ok=True)
+
+        with gw.config.update(ref_res=ref_res):
+            with gw.open(reference_image, resampling=resampling) as src:
+                profile = {
+                    'crs': src.crs,
+                    'transform': src.gw.transform,
+                    'height': src.gw.nrows,
+                    'width': src.gw.ncols,
+                    # distance (+1) + edge (+1) + crop (+1) crop types (+N)
+                    # `num_classes` includes background
+                    'count': 3 + num_classes - 1,
+                    'dtype': 'uint16',
+                    'blockxsize': 64 if 64 < src.gw.ncols else src.gw.ncols,
+                    'blockysize': 64 if 64 < src.gw.nrows else src.gw.nrows,
+                    'driver': 'GTiff',
+                    'sharing': False,
+                    'compress': compression
+                }
+        profile['tiled'] = True if max(profile['blockxsize'], profile['blockysize']) >= 16 else False
```

Comment:
The `True if ... else False` part isn't strictly needed. if clarity is the goal, probably better to wrap this expression in a function:

```python
def is_tiled(blockxsize, blockysize, tile_limit=16):
    return max(blockxsize, blockysize) >= tile_limit
```